header/p2p/session.go

```diff
@@ -0,0 +1,212 @@
+package p2p
+
+import (
+	"context"
+	"errors"
+	"io"
+	"sort"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	"github.com/celestiaorg/go-libp2p-messenger/serde"
+
+	"github.com/celestiaorg/celestia-node/header"
+	p2p_pb "github.com/celestiaorg/celestia-node/header/p2p/pb"
+)
+
+const (
+	// headersPerPeer is a maximum amount of headers that will be requested per peer
+	headersPerPeer uint64 = 64
+)
+
+// session aims to divide a range of headers
+// into several smaller requests among different peers.
+type session struct {
+	ctx        context.Context
+	host       host.Host
+	protocolID protocol.ID
+	// peerTracker contains discovered peer with records that describes peer's activity.
+	queue *peerQueue
+
+	reqCh chan *p2p_pb.ExtendedHeaderRequest
+	errCh chan error
+}
+
+func newSession(ctx context.Context, h host.Host, peerTracker []*peerStat, protocolID protocol.ID) *session {
+	return &session{
+		ctx:        ctx,
+		protocolID: protocolID,
+		host:       h,
+		queue:      newPeerQueue(peerTracker),
+		errCh:      make(chan error),
+	}
+}
+
+// doRequest chooses the best peer to fetch headers and sends a request in range of available maxRetryAttempts
+func (s *session) doRequest(
+	stat *peerStat,
+	req *p2p_pb.ExtendedHeaderRequest,
+	headers chan []*header.ExtendedHeader,
+) {
+	r, size, duration, err := s.requestHeaders(s.ctx, stat.peerID, req)
+	if err != nil {
+		if err == context.Canceled {
+			return
+		}
+		log.Errorw("requesting headers from peer failed."+
+			"Retrying the request from different peer", "failed peer", stat.peerID, "err", err)
+		s.reqCh <- req
+		return
+	}
+
+	h, err := s.processResponse(r)
+	if err != nil {
+		s.errCh <- err
+		return
+	}
+	log.Debugw("request headers from peer succeed", "from", s.host.ID(), "pID", stat.peerID, "amount", req.Amount)
+	// send headers to the channel, update peer stats and return peer to the queue, so it can be re-used in case
+	// if there are other requests awaiting
+	headers <- h
+	stat.updateStats(size, duration)
+	s.queue.push(stat)
+}
+
+// handeOutgoingRequest pops peer from the queue and sends a prepared request to the peer.
+func (s *session) handeOutgoingRequest(ctx context.Context, result chan []*header.ExtendedHeader) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-s.ctx.Done():
+			return
+		case req := <-s.reqCh:
+			stats := s.queue.calculateBestPeer()
+			go s.doRequest(stats, req, result)
+		}
```

Comment:
Is there a way here to return from for loop after we spawn all `doRequest` routines and they are done? I'm trying to figure if it could potentially be leak goroutine here. 
Maybe we can pass requests as a func param rather than through channel? Like:
```
s.handleOutgoingRequest(ctx,requests, result)
```
It could help to know when we processed last request.

Reply:
how we can know if all request had been processed if some of them could fail and we have to do request one more time?
re routines leak: I'm do not see right now but will take a deep look into this logic

Reply:
As discussed, we can enrich one of the contexts with predefined deadline so we have clear pass to return from this for loop in case requests taking too long